### PR TITLE
Fix torch.norm(dim=-1) shape for size-1 dimensions under torch.compile

### LINKED ISSUE #183121: torch.compile returns wrong shape for torch.norm(dim=-1) when reduced dimension has size 1
### 🐛 Describe the bug

`torch.compile` with the Inductor backend returns an incorrect output shape for `torch.norm(dim=-1)` when the reduced last dimension has size 1.

Eager execution correctly removes the reduced dimension and returns shape `(5, 7)`, while the compiled function returns shape `(5, 7, 1)`.

```
import torch
import torch.nn as nn

class M(nn.Module):
    def forward(self, x):
        return x.norm(dim=-1)

m = M()
torch._dynamo.reset()
c = torch.compile(m)

x = torch.rand(5, 7, 1)

es = m(x).shape
cs = c(x).shape

print(f"eager   shape = {tuple(es)}")
print(f"compile shape = {tuple(cs)}")
print("BUG" if es != cs else "OK")
```

### Error logs

eager   shape = (5, 7)
compile shape = (5, 7, 1)
BUG

### Versions

PyTorch version: 2.11.0+cu130
Is debug build: False
CUDA used to build PyTorch: 13.0
ROCM used to build PyTorch: N/A

OS: Ubuntu 24.04.2 LTS (x86_64)
GCC version: (Ubuntu 13.3.0-6ubuntu2~24.04.1) 13.3.0
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.39

Python version: 3.10.20 (main, Mar 11 2026, 17:46:40) [GCC 14.3.0] (64-bit runtime)
Python platform: Linux-5.15.167.4-microsoft-standard-WSL2-x86_64-with-glibc2.39
Is CUDA available: False
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 3080 Laptop GPU
Nvidia driver version: 545.92
cuDNN version: Could not collect
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True
Caching allocator config: N/A

Versions of relevant libraries:
[pip3] numpy==2.2.6
[pip3] onnx==1.21.0
[pip3] onnx2torch==1.5.15
[pip3] onnxruntime==1.23.2
[pip3] torch==2.11.0
[pip3] torchvision==0.26.0
[pip3] triton==3.6.0

cc @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @aakhundov @coconutruben @jataylo

### PR DESCRIPTION
**Pull request by tanpuresiddhant**
Summary:
`torch.norm(dim=-1)` returns an incorrect output shape when the reduced last dimension has size 1 and the model is compiled with the Inductor backend. Eager execution correctly removes the reduced dimension (e.g., `(5, 7, 1)` -> `(5, 7)`), while the compiled function retains it (`(5, 7, 1)`).

The bug is in `vector_norm`'s size-1 dimension optimization path in `torch/_refs/linalg/__init__.py`. When building the output shape, the code filters dimensions using `d not in dim` where `d` comes from `enumerate` (always non-negative: 0, 1, 2, ...) but `dim` can contain negative values (e.g., `[-1]`). So `2 not in [-1]` evaluates to `True` and the dimension is incorrectly kept.

The fix canonicalizes negative dim values to non-negative using modulo before the set membership check.

Closes gh-183121 Issue: https://github.com/pytorch/pytorch/issues/183121

Test Plan:
buck2 test fbcode//caffe2/test/inductor:cpu_repro_cpu -- test_vector_norm_compile test_vector_norm_negative_dim_size_one
Result: Pass 2. Fail 0.

Both `test_vector_norm_compile` (existing) and `test_vector_norm_negative_dim_size_one` (new regression test) passed.

Differential Revision: D104615065

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo

### CODE REVIEW

**Comment by ngimel** [comment]
fixed in https://github.com/pytorch/pytorch/pull/182405
